<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Steffen Durinck, Wolfgang Huber, Mike Smith" />


<title>The biomaRt users guide</title>

<link href="data:text/css;charset=utf-8,pre%20%2Eoperator%2C%0Apre%20%2Eparen%20%7B%0Acolor%3A%20rgb%28104%2C%20118%2C%20135%29%0A%7D%0Apre%20%2Eliteral%20%7B%0Acolor%3A%20%23990073%0A%7D%0Apre%20%2Enumber%20%7B%0Acolor%3A%20%23099%3B%0A%7D%0Apre%20%2Ecomment%20%7B%0Acolor%3A%20%23998%3B%0Afont%2Dstyle%3A%20italic%0A%7D%0Apre%20%2Ekeyword%20%7B%0Acolor%3A%20%23900%3B%0Afont%2Dweight%3A%20bold%0A%7D%0Apre%20%2Eidentifier%20%7B%0Acolor%3A%20rgb%280%2C%200%2C%200%29%3B%0A%7D%0Apre%20%2Estring%20%7B%0Acolor%3A%20%23d14%3B%0A%7D%0A" rel="stylesheet" />
<script src="data:application/x-javascript;base64,var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.bash=function(){var e={"true":1,"false":1};var b={cN:"variable",b:"\\$([a-zA-Z0-9_]+)\\b"};var a={cN:"variable",b:"\\$\\{(([^}])|(\\\\}))+\\}",c:[hljs.CNM]};var f={cN:"string",b:'"',e:'"',i:"\\n",c:[hljs.BE,b,a],r:0};var c={cN:"string",b:"'",e:"'",c:[{b:"''"}],r:0};var d={cN:"test_condition",b:"",e:"",c:[f,c,b,a,hljs.CNM],k:{literal:e},r:0};return{dM:{k:{keyword:{"if":1,then:1,"else":1,fi:1,"for":1,"break":1,"continue":1,"while":1,"in":1,"do":1,done:1,echo:1,exit:1,"return":1,set:1,declare:1},literal:e},c:[{cN:"shebang",b:"(#!\\/bin\\/bash)|(#!\\/bin\\/sh)",r:10},b,a,hljs.HCM,hljs.CNM,f,c,hljs.inherit(d,{b:"\\[ ",e:" \\]",r:0}),hljs.inherit(d,{b:"\\[\\[ ",e:" \\]\\]"})]}}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.css=function(){var a={cN:"function",b:hljs.IR+"\\(",e:"\\)",c:[{eW:true,eE:true,c:[hljs.NM,hljs.ASM,hljs.QSM]}]};return{cI:true,dM:{i:"[=/|']",c:[hljs.CBLCLM,{cN:"id",b:"\\#[A-Za-z0-9_-]+"},{cN:"class",b:"\\.[A-Za-z0-9_-]+",r:0},{cN:"attr_selector",b:"\\[",e:"\\]",i:"$"},{cN:"pseudo",b:":(:)?[a-zA-Z0-9\\_\\-\\+\\(\\)\\\"\\']+"},{cN:"at_rule",b:"@(font-face|page)",l:"[a-z-]+",k:{"font-face":1,page:1}},{cN:"at_rule",b:"@",e:"[{;]",eE:true,k:{"import":1,page:1,media:1,charset:1},c:[a,hljs.ASM,hljs.QSM,hljs.NM]},{cN:"tag",b:hljs.IR,r:0},{cN:"rules",b:"{",e:"}",i:"[^\\s]",r:0,c:[hljs.CBLCLM,{cN:"rule",b:"[^\\s]",rB:true,e:";",eW:true,c:[{cN:"attribute",b:"[A-Z\\_\\.\\-]+",e:":",eE:true,i:"[^\\s]",starts:{cN:"value",eW:true,eE:true,c:[a,hljs.NM,hljs.QSM,hljs.ASM,hljs.CBLCLM,{cN:"hexcolor",b:"\\#[0-9A-F]+"},{cN:"important",b:"!important"}]}}]}]}]}}}();hljs.LANGUAGES.ini={cI:true,dM:{i:"[^\\s]",c:[{cN:"comment",b:";",e:"$"},{cN:"title",b:"^\\[",e:"\\]"},{cN:"setting",b:"^[a-z0-9_\\[\\]]+[ \\t]*=[ \\t]*",e:"$",c:[{cN:"value",eW:true,k:{on:1,off:1,"true":1,"false":1,yes:1,no:1},c:[hljs.QSM,hljs.NM]}]}]}};hljs.LANGUAGES.perl=function(){var d={getpwent:1,getservent:1,quotemeta:1,msgrcv:1,scalar:1,kill:1,dbmclose:1,undef:1,lc:1,ma:1,syswrite:1,tr:1,send:1,umask:1,sysopen:1,shmwrite:1,vec:1,qx:1,utime:1,local:1,oct:1,semctl:1,localtime:1,readpipe:1,"do":1,"return":1,format:1,read:1,sprintf:1,dbmopen:1,pop:1,getpgrp:1,not:1,getpwnam:1,rewinddir:1,qq:1,fileno:1,qw:1,endprotoent:1,wait:1,sethostent:1,bless:1,s:0,opendir:1,"continue":1,each:1,sleep:1,endgrent:1,shutdown:1,dump:1,chomp:1,connect:1,getsockname:1,die:1,socketpair:1,close:1,flock:1,exists:1,index:1,shmget:1,sub:1,"for":1,endpwent:1,redo:1,lstat:1,msgctl:1,setpgrp:1,abs:1,exit:1,select:1,print:1,ref:1,gethostbyaddr:1,unshift:1,fcntl:1,syscall:1,"goto":1,getnetbyaddr:1,join:1,gmtime:1,symlink:1,semget:1,splice:1,x:0,getpeername:1,recv:1,log:1,setsockopt:1,cos:1,last:1,reverse:1,gethostbyname:1,getgrnam:1,study:1,formline:1,endhostent:1,times:1,chop:1,length:1,gethostent:1,getnetent:1,pack:1,getprotoent:1,getservbyname:1,rand:1,mkdir:1,pos:1,chmod:1,y:0,substr:1,endnetent:1,printf:1,next:1,open:1,msgsnd:1,readdir:1,use:1,unlink:1,getsockopt:1,getpriority:1,rindex:1,wantarray:1,hex:1,system:1,getservbyport:1,endservent:1,"int":1,chr:1,untie:1,rmdir:1,prototype:1,tell:1,listen:1,fork:1,shmread:1,ucfirst:1,setprotoent:1,"else":1,sysseek:1,link:1,getgrgid:1,shmctl:1,waitpid:1,unpack:1,getnetbyname:1,reset:1,chdir:1,grep:1,split:1,require:1,caller:1,lcfirst:1,until:1,warn:1,"while":1,values:1,shift:1,telldir:1,getpwuid:1,my:1,getprotobynumber:1,"delete":1,and:1,sort:1,uc:1,defined:1,srand:1,accept:1,"package":1,seekdir:1,getprotobyname:1,semop:1,our:1,rename:1,seek:1,"if":1,q:0,chroot:1,sysread:1,setpwent:1,no:1,crypt:1,getc:1,chown:1,sqrt:1,write:1,setnetent:1,setpriority:1,foreach:1,tie:1,sin:1,msgget:1,map:1,stat:1,getlogin:1,unless:1,elsif:1,truncate:1,exec:1,keys:1,glob:1,tied:1,closedir:1,ioctl:1,socket:1,readlink:1,"eval":1,xor:1,readline:1,binmode:1,setservent:1,eof:1,ord:1,bind:1,alarm:1,pipe:1,atan2:1,getgrent:1,exp:1,time:1,push:1,setgrent:1,gt:1,lt:1,or:1,ne:1,m:0};var f={cN:"subst",b:"[$@]\\{",e:"\\}",k:d,r:10};var c={cN:"variable",b:"\\$\\d"};var b={cN:"variable",b:"[\\$\\%\\@\\*](\\^\\w\\b|#\\w+(\\:\\:\\w+)*|[^\\s\\w{]|{\\w+}|\\w+(\\:\\:\\w*)*)"};var h=[hljs.BE,f,c,b];var g={b:"->",c:[{b:hljs.IR},{b:"{",e:"}"}]};var e={cN:"comment",b:"^(__END__|__DATA__)",e:"\\n$",r:5};var a=[c,b,hljs.HCM,e,g,{cN:"string",b:"q[qwxr]?\\s*\\(",e:"\\)",c:h,r:5},{cN:"string",b:"q[qwxr]?\\s*\\[",e:"\\]",c:h,r:5},{cN:"string",b:"q[qwxr]?\\s*\\{",e:"\\}",c:h,r:5},{cN:"string",b:"q[qwxr]?\\s*\\|",e:"\\|",c:h,r:5},{cN:"string",b:"q[qwxr]?\\s*\\<",e:"\\>",c:h,r:5},{cN:"string",b:"qw\\s+q",e:"q",c:h,r:5},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"string",b:'"',e:'"',c:h,r:0},{cN:"string",b:"`",e:"`",c:[hljs.BE]},{cN:"string",b:"{\\w+}",r:0},{cN:"string",b:"-?\\w+\\s*\\=\\>",r:0},{cN:"number",b:"(\\b0[0-7_]+)|(\\b0x[0-9a-fA-F_]+)|(\\b[1-9][0-9_]*(\\.[0-9_]+)?)|[0_]\\b",r:0},{b:"("+hljs.RSR+"|\\b(split|return|print|reverse|grep)\\b)\\s*",k:{split:1,"return":1,print:1,reverse:1,grep:1},r:0,c:[hljs.HCM,e,{cN:"regexp",b:"(s|tr|y)/(\\\\.|[^/])*/(\\\\.|[^/])*/[a-z]*",r:10},{cN:"regexp",b:"(m|qr)?/",e:"/[a-z]*",c:[hljs.BE],r:0}]},{cN:"sub",b:"\\bsub\\b",e:"(\\s*\\(.*?\\))?[;{]",k:{sub:1},r:5},{cN:"operator",b:"-\\w\\b",r:0},{cN:"pod",b:"\\=\\w",e:"\\=cut"}];f.c=a;g.c[1].c=a;return{dM:{k:d,c:a}}}();hljs.LANGUAGES.python=function(){var b=[{cN:"string",b:"(u|b)?r?'''",e:"'''",r:10},{cN:"string",b:'(u|b)?r?"""',e:'"""',r:10},{cN:"string",b:"(u|r|ur)'",e:"'",c:[hljs.BE],r:10},{cN:"string",b:'(u|r|ur)"',e:'"',c:[hljs.BE],r:10},{cN:"string",b:"(b|br)'",e:"'",c:[hljs.BE]},{cN:"string",b:'(b|br)"',e:'"',c:[hljs.BE]}].concat([hljs.ASM,hljs.QSM]);var d={cN:"title",b:hljs.UIR};var c={cN:"params",b:"\\(",e:"\\)",c:b.concat([hljs.CNM])};var a={bWK:true,e:":",i:"[${]",c:[d,c],r:10};return{dM:{k:{keyword:{and:1,elif:1,is:1,global:1,as:1,"in":1,"if":1,from:1,raise:1,"for":1,except:1,"finally":1,print:1,"import":1,pass:1,"return":1,exec:1,"else":1,"break":1,not:1,"with":1,"class":1,assert:1,yield:1,"try":1,"while":1,"continue":1,del:1,or:1,def:1,lambda:1,nonlocal:10},built_in:{None:1,True:1,False:1,Ellipsis:1,NotImplemented:1}},i:"(</|->|\\?)",c:b.concat([hljs.HCM,hljs.inherit(a,{cN:"function",k:{def:1}}),hljs.inherit(a,{cN:"class",k:{"class":1}}),hljs.CNM,{cN:"decorator",b:"@",e:"$"}])}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};hljs.LANGUAGES.ruby=function(){var a="[a-zA-Z_][a-zA-Z0-9_]*(\\!|\\?)?";var j="[a-zA-Z_]\\w*[!?=]?|[-+~]\\@|<<|>>|=~|===?|<=>|[<>]=?|\\*\\*|[-/+%^&*~`|]|\\[\\]=?";var f={keyword:{and:1,"false":1,then:1,defined:1,module:1,"in":1,"return":1,redo:1,"if":1,BEGIN:1,retry:1,end:1,"for":1,"true":1,self:1,when:1,next:1,until:1,"do":1,begin:1,unless:1,END:1,rescue:1,nil:1,"else":1,"break":1,undef:1,not:1,"super":1,"class":1,"case":1,require:1,yield:1,alias:1,"while":1,ensure:1,elsif:1,or:1,def:1},keymethods:{__id__:1,__send__:1,abort:1,abs:1,"all?":1,allocate:1,ancestors:1,"any?":1,arity:1,assoc:1,at:1,at_exit:1,autoload:1,"autoload?":1,"between?":1,binding:1,binmode:1,"block_given?":1,call:1,callcc:1,caller:1,capitalize:1,"capitalize!":1,casecmp:1,"catch":1,ceil:1,center:1,chomp:1,"chomp!":1,chop:1,"chop!":1,chr:1,"class":1,class_eval:1,"class_variable_defined?":1,class_variables:1,clear:1,clone:1,close:1,close_read:1,close_write:1,"closed?":1,coerce:1,collect:1,"collect!":1,compact:1,"compact!":1,concat:1,"const_defined?":1,const_get:1,const_missing:1,const_set:1,constants:1,count:1,crypt:1,"default":1,default_proc:1,"delete":1,"delete!":1,delete_at:1,delete_if:1,detect:1,display:1,div:1,divmod:1,downcase:1,"downcase!":1,downto:1,dump:1,dup:1,each:1,each_byte:1,each_index:1,each_key:1,each_line:1,each_pair:1,each_value:1,each_with_index:1,"empty?":1,entries:1,eof:1,"eof?":1,"eql?":1,"equal?":1,"eval":1,exec:1,exit:1,"exit!":1,extend:1,fail:1,fcntl:1,fetch:1,fileno:1,fill:1,find:1,find_all:1,first:1,flatten:1,"flatten!":1,floor:1,flush:1,for_fd:1,foreach:1,fork:1,format:1,freeze:1,"frozen?":1,fsync:1,getc:1,gets:1,global_variables:1,grep:1,gsub:1,"gsub!":1,"has_key?":1,"has_value?":1,hash:1,hex:1,id:1,include:1,"include?":1,included_modules:1,index:1,indexes:1,indices:1,induced_from:1,inject:1,insert:1,inspect:1,instance_eval:1,instance_method:1,instance_methods:1,"instance_of?":1,"instance_variable_defined?":1,instance_variable_get:1,instance_variable_set:1,instance_variables:1,"integer?":1,intern:1,invert:1,ioctl:1,"is_a?":1,isatty:1,"iterator?":1,join:1,"key?":1,keys:1,"kind_of?":1,lambda:1,last:1,length:1,lineno:1,ljust:1,load:1,local_variables:1,loop:1,lstrip:1,"lstrip!":1,map:1,"map!":1,match:1,max:1,"member?":1,merge:1,"merge!":1,method:1,"method_defined?":1,method_missing:1,methods:1,min:1,module_eval:1,modulo:1,name:1,nesting:1,"new":1,next:1,"next!":1,"nil?":1,nitems:1,"nonzero?":1,object_id:1,oct:1,open:1,pack:1,partition:1,pid:1,pipe:1,pop:1,popen:1,pos:1,prec:1,prec_f:1,prec_i:1,print:1,printf:1,private_class_method:1,private_instance_methods:1,"private_method_defined?":1,private_methods:1,proc:1,protected_instance_methods:1,"protected_method_defined?":1,protected_methods:1,public_class_method:1,public_instance_methods:1,"public_method_defined?":1,public_methods:1,push:1,putc:1,puts:1,quo:1,raise:1,rand:1,rassoc:1,read:1,read_nonblock:1,readchar:1,readline:1,readlines:1,readpartial:1,rehash:1,reject:1,"reject!":1,remainder:1,reopen:1,replace:1,require:1,"respond_to?":1,reverse:1,"reverse!":1,reverse_each:1,rewind:1,rindex:1,rjust:1,round:1,rstrip:1,"rstrip!":1,scan:1,seek:1,select:1,send:1,set_trace_func:1,shift:1,singleton_method_added:1,singleton_methods:1,size:1,sleep:1,slice:1,"slice!":1,sort:1,"sort!":1,sort_by:1,split:1,sprintf:1,squeeze:1,"squeeze!":1,srand:1,stat:1,step:1,store:1,strip:1,"strip!":1,sub:1,"sub!":1,succ:1,"succ!":1,sum:1,superclass:1,swapcase:1,"swapcase!":1,sync:1,syscall:1,sysopen:1,sysread:1,sysseek:1,system:1,syswrite:1,taint:1,"tainted?":1,tell:1,test:1,"throw":1,times:1,to_a:1,to_ary:1,to_f:1,to_hash:1,to_i:1,to_int:1,to_io:1,to_proc:1,to_s:1,to_str:1,to_sym:1,tr:1,"tr!":1,tr_s:1,"tr_s!":1,trace_var:1,transpose:1,trap:1,truncate:1,"tty?":1,type:1,ungetc:1,uniq:1,"uniq!":1,unpack:1,unshift:1,untaint:1,untrace_var:1,upcase:1,"upcase!":1,update:1,upto:1,"value?":1,values:1,values_at:1,warn:1,write:1,write_nonblock:1,"zero?":1,zip:1}};var c={cN:"yardoctag",b:"@[A-Za-z]+"};var k=[{cN:"comment",b:"#",e:"$",c:[c]},{cN:"comment",b:"^\\=begin",e:"^\\=end",c:[c],r:10},{cN:"comment",b:"^__END__",e:"\\n$"}];var d={cN:"subst",b:"#\\{",e:"}",l:a,k:f};var i=[hljs.BE,d];var b=[{cN:"string",b:"'",e:"'",c:i,r:0},{cN:"string",b:'"',e:'"',c:i,r:0},{cN:"string",b:"%[qw]?\\(",e:"\\)",c:i,r:10},{cN:"string",b:"%[qw]?\\[",e:"\\]",c:i,r:10},{cN:"string",b:"%[qw]?{",e:"}",c:i,r:10},{cN:"string",b:"%[qw]?<",e:">",c:i,r:10},{cN:"string",b:"%[qw]?/",e:"/",c:i,r:10},{cN:"string",b:"%[qw]?%",e:"%",c:i,r:10},{cN:"string",b:"%[qw]?-",e:"-",c:i,r:10},{cN:"string",b:"%[qw]?\\|",e:"\\|",c:i,r:10}];var h={cN:"function",b:"\\bdef\\s+",e:" |$|;",l:a,k:f,c:[{cN:"title",b:j,l:a,k:f},{cN:"params",b:"\\(",e:"\\)",l:a,k:f}].concat(k)};var g={cN:"identifier",b:a,l:a,k:f,r:0};var e=k.concat(b.concat([{cN:"class",b:"\\b(class|module)\\b",e:"$|;",k:{"class":1,module:1},c:[{cN:"title",b:"[A-Za-z_]\\w*(::\\w+)*(\\?|\\!)?",r:0},{cN:"inheritance",b:"<\\s*",c:[{cN:"parent",b:"("+hljs.IR+"::)?"+hljs.IR}]}].concat(k)},h,{cN:"constant",b:"(::)?([A-Z]\\w*(::)?)+",r:0},{cN:"symbol",b:":",c:b.concat([g]),r:0},{cN:"number",b:"(\\b0[0-7_]+)|(\\b0x[0-9a-fA-F_]+)|(\\b[1-9][0-9_]*(\\.[0-9_]+)?)|[0_]\\b",r:0},{cN:"number",b:"\\?\\w"},{cN:"variable",b:"(\\$\\W)|((\\$|\\@\\@?)(\\w+))"},g,{b:"("+hljs.RSR+")\\s*",c:k.concat([{cN:"regexp",b:"/",e:"/[a-z]*",i:"\\n",c:[hljs.BE]}]),r:0}]));d.c=e;h.c[1].c=e;return{dM:{l:a,k:f,c:e}}}();hljs.LANGUAGES.scala=function(){var b={cN:"annotation",b:"@[A-Za-z]+"};var a={cN:"string",b:'u?r?"""',e:'"""',r:10};return{dM:{k:{type:1,yield:1,lazy:1,override:1,def:1,"with":1,val:1,"var":1,"false":1,"true":1,sealed:1,"abstract":1,"private":1,trait:1,object:1,"null":1,"if":1,"for":1,"while":1,"throw":1,"finally":1,"protected":1,"extends":1,"import":1,"final":1,"return":1,"else":1,"break":1,"new":1,"catch":1,"super":1,"class":1,"case":1,"package":1,"default":1,"try":1,"this":1,match:1,"continue":1,"throws":1},c:[{cN:"javadoc",b:"/\\*\\*",e:"\\*/",c:[{cN:"javadoctag",b:"@[A-Za-z]+"}],r:10},hljs.CLCM,hljs.CBLCLM,hljs.ASM,hljs.QSM,a,{cN:"class",b:"((case )?class |object |trait )",e:"({|$)",i:":",k:{"case":1,"class":1,trait:1,object:1},c:[{bWK:true,k:{"extends":1,"with":1},r:10},{cN:"title",b:hljs.UIR},{cN:"params",b:"\\(",e:"\\)",c:[hljs.ASM,hljs.QSM,a,b]}]},hljs.CNM,b]}}}();hljs.LANGUAGES.sql={cI:true,dM:{i:"[^\\s]",c:[{cN:"operator",b:"(begin|start|commit|rollback|savepoint|lock|alter|create|drop|rename|call|delete|do|handler|insert|load|replace|select|truncate|update|set|show|pragma|grant)\\b",e:";|"+hljs.ER,k:{keyword:{all:1,partial:1,global:1,month:1,current_timestamp:1,using:1,go:1,revoke:1,smallint:1,indicator:1,"end-exec":1,disconnect:1,zone:1,"with":1,character:1,assertion:1,to:1,add:1,current_user:1,usage:1,input:1,local:1,alter:1,match:1,collate:1,real:1,then:1,rollback:1,get:1,read:1,timestamp:1,session_user:1,not:1,integer:1,bit:1,unique:1,day:1,minute:1,desc:1,insert:1,execute:1,like:1,ilike:2,level:1,decimal:1,drop:1,"continue":1,isolation:1,found:1,where:1,constraints:1,domain:1,right:1,national:1,some:1,module:1,transaction:1,relative:1,second:1,connect:1,escape:1,close:1,system_user:1,"for":1,deferred:1,section:1,cast:1,current:1,sqlstate:1,allocate:1,intersect:1,deallocate:1,numeric:1,"public":1,preserve:1,full:1,"goto":1,initially:1,asc:1,no:1,key:1,output:1,collation:1,group:1,by:1,union:1,session:1,both:1,last:1,language:1,constraint:1,column:1,of:1,space:1,foreign:1,deferrable:1,prior:1,connection:1,unknown:1,action:1,commit:1,view:1,or:1,first:1,into:1,"float":1,year:1,primary:1,cascaded:1,except:1,restrict:1,set:1,references:1,names:1,table:1,outer:1,open:1,select:1,size:1,are:1,rows:1,from:1,prepare:1,distinct:1,leading:1,create:1,only:1,next:1,inner:1,authorization:1,schema:1,corresponding:1,option:1,declare:1,precision:1,immediate:1,"else":1,timezone_minute:1,external:1,varying:1,translation:1,"true":1,"case":1,exception:1,join:1,hour:1,"default":1,"double":1,scroll:1,value:1,cursor:1,descriptor:1,values:1,dec:1,fetch:1,procedure:1,"delete":1,and:1,"false":1,"int":1,is:1,describe:1,"char":1,as:1,at:1,"in":1,varchar:1,"null":1,trailing:1,any:1,absolute:1,current_time:1,end:1,grant:1,privileges:1,when:1,cross:1,check:1,write:1,current_date:1,pad:1,begin:1,temporary:1,exec:1,time:1,update:1,catalog:1,user:1,sql:1,date:1,on:1,identity:1,timezone_hour:1,natural:1,whenever:1,interval:1,work:1,order:1,cascade:1,diagnostics:1,nchar:1,having:1,left:1,call:1,"do":1,handler:1,load:1,replace:1,truncate:1,start:1,lock:1,show:1,pragma:1},aggregate:{count:1,sum:1,min:1,max:1,avg:1}},c:[{cN:"string",b:"'",e:"'",c:[hljs.BE,{b:"''"}],r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE,{b:'""'}],r:0},{cN:"string",b:"`",e:"`",c:[hljs.BE]},hljs.CNM]},hljs.CBLCLM,{cN:"comment",b:"--",e:"$"}]}};hljs.LANGUAGES.stan={dM:{c:[hljs.HCM,hljs.CLCM,hljs.QSM,hljs.CNM,{cN:"operator",b:"(?:<-|~|\\|\\||&&|==|!=|<=?|>=?|\\+|-|\\.?/|\\\\|\\^|\\^|!|'|%|:|,|;|=)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0},{cN:"function",b:"(?:Phi|Phi_approx|abs|acos|acosh|append_col|append_row|asin|asinh|atan|atan2|atanh|bernoulli_ccdf_log|bernoulli_cdf|bernoulli_cdf_log|bernoulli_log|bernoulli_logit_log|bernoulli_rng|bessel_first_kind|bessel_second_kind|beta_binomial_ccdf_log|beta_binomial_cdf|beta_binomial_cdf_log|beta_binomial_log|beta_binomial_rng|beta_ccdf_log|beta_cdf|beta_cdf_log|beta_log|beta_rng|binary_log_loss|binomial_ccdf_log|binomial_cdf|binomial_cdf_log|binomial_coefficient_log|binomial_log|binomial_logit_log|binomial_rng|block|categorical_log|categorical_logit_log|categorical_rng|cauchy_ccdf_log|cauchy_cdf|cauchy_cdf_log|cauchy_log|cauchy_rng|cbrt|ceil|chi_square_ccdf_log|chi_square_cdf|chi_square_cdf_log|chi_square_log|chi_square_rng|cholesky_decompose|col|cols|columns_dot_product|columns_dot_self|cos|cosh|crossprod|csr_extract_u|csr_extract_v|csr_extract_w|csr_matrix_times_vector|csr_to_dense_matrix|cumulative_sum|determinant|diag_matrix|diag_post_multiply|diag_pre_multiply|diagonal|digamma|dims|dirichlet_log|dirichlet_rng|distance|dot_product|dot_self|double_exponential_ccdf_log|double_exponential_cdf|double_exponential_cdf_log|double_exponential_log|double_exponential_rng|e|eigenvalues_sym|eigenvectors_sym|erf|erfc|exp|exp2|exp_mod_normal_ccdf_log|exp_mod_normal_cdf|exp_mod_normal_cdf_log|exp_mod_normal_log|exp_mod_normal_rng|expm1|exponential_ccdf_log|exponential_cdf|exponential_cdf_log|exponential_log|exponential_rng|fabs|falling_factorial|fdim|floor|fma|fmax|fmin|fmod|frechet_ccdf_log|frechet_cdf|frechet_cdf_log|frechet_log|frechet_rng|gamma_ccdf_log|gamma_cdf|gamma_cdf_log|gamma_log|gamma_p|gamma_q|gamma_rng|gaussian_dlm_obs_log|get_lp|gumbel_ccdf_log|gumbel_cdf|gumbel_cdf_log|gumbel_log|gumbel_rng|head|hypergeometric_log|hypergeometric_rng|hypot|if_else|int_step|inv|inv_chi_square_ccdf_log|inv_chi_square_cdf|inv_chi_square_cdf_log|inv_chi_square_log|inv_chi_square_rng|inv_cloglog|inv_gamma_ccdf_log|inv_gamma_cdf|inv_gamma_cdf_log|inv_gamma_log|inv_gamma_rng|inv_logit|inv_phi|inv_sqrt|inv_square|inv_wishart_log|inv_wishart_rng|inverse|inverse_spd|is_inf|is_nan|lbeta|lgamma|lkj_corr_cholesky_log|lkj_corr_cholesky_rng|lkj_corr_log|lkj_corr_rng|lmgamma|log|log10|log1m|log1m_exp|log1m_inv_logit|log1p|log1p_exp|log2|log_determinant|log_diff_exp|log_falling_factorial|log_inv_logit|log_mix|log_rising_factorial|log_softmax|log_sum_exp|logistic_ccdf_log|logistic_cdf|logistic_cdf_log|logistic_log|logistic_rng|logit|lognormal_ccdf_log|lognormal_cdf|lognormal_cdf_log|lognormal_log|lognormal_rng|machine_precision|max|mdivide_left_tri_low|mdivide_right_tri_low|mean|min|modified_bessel_first_kind|modified_bessel_second_kind|multi_gp_cholesky_log|multi_gp_log|multi_normal_cholesky_log|multi_normal_cholesky_rng|multi_normal_log|multi_normal_prec_log|multi_normal_rng|multi_student_t_log|multi_student_t_rng|multinomial_log|multinomial_rng|multiply_log|multiply_lower_tri_self_transpose|neg_binomial_2_ccdf_log|neg_binomial_2_cdf|neg_binomial_2_cdf_log|neg_binomial_2_log|neg_binomial_2_log_log|neg_binomial_2_log_rng|neg_binomial_2_rng|neg_binomial_ccdf_log|neg_binomial_cdf|neg_binomial_cdf_log|neg_binomial_log|neg_binomial_rng|negative_infinity|normal_ccdf_log|normal_cdf|normal_cdf_log|normal_log|normal_rng|not_a_number|num_elements|ordered_logistic_log|ordered_logistic_rng|owens_t|pareto_ccdf_log|pareto_cdf|pareto_cdf_log|pareto_log|pareto_rng|pareto_type_2_ccdf_log|pareto_type_2_cdf|pareto_type_2_cdf_log|pareto_type_2_log|pareto_type_2_rng|pi|poisson_ccdf_log|poisson_cdf|poisson_cdf_log|poisson_log|poisson_log_log|poisson_log_rng|poisson_rng|positive_infinity|pow|prod|qr_Q|qr_R|quad_form|quad_form_diag|quad_form_sym|rank|rayleigh_ccdf_log|rayleigh_cdf|rayleigh_cdf_log|rayleigh_log|rayleigh_rng|rep_array|rep_matrix|rep_row_vector|rep_vector|rising_factorial|round|row|rows|rows_dot_product|rows_dot_self|scaled_inv_chi_square_ccdf_log|scaled_inv_chi_square_cdf|scaled_inv_chi_square_cdf_log|scaled_inv_chi_square_log|scaled_inv_chi_square_rng|sd|segment|sin|singular_values|sinh|size|skew_normal_ccdf_log|skew_normal_cdf|skew_normal_cdf_log|skew_normal_log|skew_normal_rng|softmax|sort_asc|sort_desc|sort_indices_asc|sort_indices_desc|sqrt|sqrt2|square|squared_distance|step|student_t_ccdf_log|student_t_cdf|student_t_cdf_log|student_t_log|student_t_rng|sub_col|sub_row|sum|tail|tan|tanh|tcrossprod|tgamma|to_array_1d|to_array_2d|to_matrix|to_row_vector|to_vector|trace|trace_gen_quad_form|trace_quad_form|trigamma|trunc|uniform_ccdf_log|uniform_cdf|uniform_cdf_log|uniform_log|uniform_rng|variance|von_mises_log|von_mises_rng|weibull_ccdf_log|weibull_cdf|weibull_cdf_log|weibull_log|weibull_rng|wiener_log|wishart_log|wishart_rng)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"function",b:"(?:bernoulli|bernoulli_logit|beta|beta_binomial|binomial|binomial_logit|categorical|categorical_logit|cauchy|chi_square|dirichlet|double_exponential|exp_mod_normal|exponential|frechet|gamma|gaussian_dlm_obs|gumbel|hypergeometric|inv_chi_square|inv_gamma|inv_wishart|lkj_corr|lkj_corr_cholesky|logistic|lognormal|multi_gp|multi_gp_cholesky|multi_normal|multi_normal_cholesky|multi_normal_prec|multi_student_t|multinomial|neg_binomial|neg_binomial_2|neg_binomial_2_log|normal|ordered_logistic|pareto|pareto_type_2|poisson|poisson_log|rayleigh|scaled_inv_chi_square|skew_normal|student_t|uniform|von_mises|weibull|wiener|wishart)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"(?:for|in|while|if|then|else|return|lower|upper|print|increment_log_prob|integrate_ode|reject)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"(?:int|real|vector|simplex|unit_vector|ordered|positive_ordered|row_vector|matrix|cholesky_factor_cov|cholesky_factor_corr|corr_matrix|cov_matrix|void)\\b",e:hljs.IMMEDIATE_RE,r:5},{cN:"keyword",b:"(?:functions|data|transformed\\s+data|parameters|transformed\\s+parameters|model|generated\\s+quantities)\\b",e:hljs.IMMEDIATE_RE,r:5}]}};hljs.LANGUAGES.xml=function(){var b="[A-Za-z0-9\\._:-]+";var a={eW:true,c:[{cN:"attribute",b:b,r:0},{b:'="',rB:true,e:'"',c:[{cN:"value",b:'"',eW:true}]},{b:"='",rB:true,e:"'",c:[{cN:"value",b:"'",eW:true}]},{b:"=",c:[{cN:"value",b:"[^\\s/>]+"}]}]};return{cI:true,dM:{c:[{cN:"pi",b:"<\\?",e:"\\?>",r:10},{cN:"doctype",b:"<!DOCTYPE",e:">",r:10,c:[{b:"\\[",e:"\\]"}]},{cN:"comment",b:"<!--",e:"-->",r:10},{cN:"cdata",b:"<\\!\\[CDATA\\[",e:"\\]\\]>",r:10},{cN:"tag",b:"<style(?=\\s|>|$)",e:">",k:{title:{style:1}},c:[a],starts:{cN:"css",e:"</style>",rE:true,sL:"css"}},{cN:"tag",b:"<script(?=\\s|>|$)",e:">",k:{title:{script:1}},c:[a],starts:{cN:"javascript",e:"<\/script>",rE:true,sL:"javascript"}},{cN:"vbscript",b:"<%",e:"%>",sL:"vbscript"},{cN:"tag",b:"</?",e:"/?>",c:[{cN:"title",b:"[^ />]+"},a]}]}}}();
hljs.initHighlightingOnLoad();

"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>


<link href="data:text/css;charset=utf-8,body%2C%20td%20%7B%0Afont%2Dfamily%3A%20sans%2Dserif%3B%0Abackground%2Dcolor%3A%20white%3B%0Afont%2Dsize%3A%2013px%3B%0A%7D%0Abody%20%7B%0Amax%2Dwidth%3A%20800px%3B%0Amargin%3A%200%20auto%3B%0Apadding%3A%201em%201em%202em%3B%0Aline%2Dheight%3A%2020px%3B%0A%7D%0A%0Adiv%23TOC%20li%20%7B%0Alist%2Dstyle%3Anone%3B%0Abackground%2Dimage%3Anone%3B%0Abackground%2Drepeat%3Anone%3B%0Abackground%2Dposition%3A0%3B%0A%7D%0A%0Ap%2C%20pre%20%7B%20margin%3A%200em%200em%201em%3B%0A%7D%0A%0Aimg%2C%20table%20%7B%0Amargin%3A%200em%20auto%201em%3B%0A%7D%0Ap%20%7B%0Atext%2Dalign%3A%20justify%3B%0A%7D%0Att%2C%20code%2C%20pre%20%7B%0Afont%2Dfamily%3A%20%27DejaVu%20Sans%20Mono%27%2C%20%27Droid%20Sans%20Mono%27%2C%20%27Lucida%20Console%27%2C%20Consolas%2C%20Monaco%2C%20monospace%3B%0A%7D%0Ah1%2C%20h2%2C%20h3%2C%20h4%2C%20h5%2C%20h6%20%7B%20font%2Dfamily%3A%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Amargin%3A%201%2E2em%200em%200%2E6em%200em%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%2Etitle%20%7B%0Afont%2Dsize%3A%20250%25%3B%0Afont%2Dweight%3A%20normal%3B%0Acolor%3A%20%2387b13f%3B%0Aline%2Dheight%3A%201%2E1em%3B%0Amargin%2Dtop%3A%200px%3B%0Aborder%2Dbottom%3A%200px%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%20160%25%3B%0Afont%2Dweight%3A%20normal%3B%0Aline%2Dheight%3A%201%2E4em%3B%0Aborder%2Dbottom%3A%201px%20%231a81c2%20solid%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%20130%25%3B%20%7D%0Ah1%2C%20h2%2C%20h3%20%7B%0Acolor%3A%20%231a81c2%3B%0A%7D%0Ah3%2C%20h4%2C%20h5%2C%20h6%20%7B%0Afont%2Dsize%3A115%25%3B%0A%7D%20%0A%0Aa%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Aactive%20%7B%20outline%3A%20none%3B%20%7D%0Aa%3Avisited%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Ahover%20%7B%20color%3A%20%234c94c2%3B%20%7D%0Apre%2C%20img%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Adisplay%3A%20block%3B%0A%7D%0Apre%20%7B%0Aborder%3A%200px%20none%3B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0Awhite%2Dspace%3A%20pre%3B%0Aoverflow%2Dx%3A%20auto%3B%0A%7D%0Apre%20code%20%7B%0Aborder%3A%201px%20%23aaa%20dashed%3B%0Abackground%2Dcolor%3A%20white%3B%0Adisplay%3A%20block%3B%0Apadding%3A%201em%3B%20color%3A%20%23111%3B%0Aoverflow%2Dx%3A%20inherit%3B%0A%7D%0A%0Apre%20code%5Bclass%5D%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Apre%5Bclass%5D%20code%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Acode%20%7B%20background%2Dcolor%3A%20transparent%3B%0Acolor%3A%20%2387b13f%3B%0Afont%2Dsize%3A%2092%25%3B%0A%7D%0A%0Atable%2C%20td%2C%20th%20%7B%0Aborder%3A%20none%3B%0Apadding%3A%200%200%2E5em%3B%0A%7D%0A%0Atbody%20tr%3Anth%2Dchild%28odd%29%20td%20%7B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0A%7D%0Ablockquote%20%7B%0Acolor%3A%23666666%3B%0Amargin%3A0%3B%0Apadding%2Dleft%3A%201em%3B%0Aborder%2Dleft%3A%200%2E5em%20%23EEE%20solid%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200px%3B%0Aborder%2Dbottom%3A%20none%3B%0Aborder%2Dtop%2Dwidth%3A%20thin%3B%0Aborder%2Dtop%2Dstyle%3A%20dotted%3B%0Aborder%2Dtop%2Dcolor%3A%20%23999999%3B%0A%7D%0Aspan%2Eheader%2Dsection%2Dnumber%20%7B%0Apadding%2Dright%3A%201em%3B%0A%7D%0Aspan%2Etoc%2Dsection%2Dnumber%3A%3Aafter%20%7B%0Acontent%3A%20%22%20%22%3B%0Awhite%2Dspace%3A%20pre%3B%0A%7D%0A%40media%20print%20%7B%0A%2A%20%7B%0Abackground%3A%20transparent%20%21important%3B%0Acolor%3A%20black%20%21important%3B%0Afilter%3Anone%20%21important%3B%0A%2Dms%2Dfilter%3A%20none%20%21important%3B%0A%7D%0Abody%20%7B%0Afont%2Dsize%3A12pt%3B%0Amax%2Dwidth%3A100%25%3B%0A%7D%0Aa%2C%20a%3Avisited%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Avisibility%3A%20hidden%3B%0Apage%2Dbreak%2Dbefore%3A%20always%3B%0A%7D%0Apre%2C%20blockquote%20%7B%0Apadding%2Dright%3A%201em%3B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Atr%2C%20img%20%7B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%20%21important%3B%0A%7D%0A%40page%20%3Aleft%20%7B%0Amargin%3A%2015mm%2020mm%2015mm%2010mm%3B%0A%7D%0A%40page%20%3Aright%20%7B%0Amargin%3A%2015mm%2010mm%2015mm%2020mm%3B%0A%7D%0Ap%2C%20h2%2C%20h3%20%7B%0Aorphans%3A%203%3B%20widows%3A%203%3B%0A%7D%0Ah2%2C%20h3%20%7B%0Apage%2Dbreak%2Dafter%3A%20avoid%3B%0A%7D%0A%7D%0A" rel="stylesheet" type="text/css" />

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if(links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>

</head>

<body>


<div id="header">
<h1 class="title">The biomaRt users guide</h1>
<h4 class="author"><em>Steffen Durinck, Wolfgang Huber, Mike Smith</em></h4>
</div>
<h4 class="package">Package version: <span style="font-weight: normal">biomaRt 2.32.1</span></h4>

<h1>Contents</h1>
<div id="TOC">
<ul>
<li><a href="#introduction"><span class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#selecting-a-biomart-database-and-dataset"><span class="toc-section-number">2</span> Selecting a BioMart database and dataset</a></li>
<li><a href="#how-to-build-a-biomart-query"><span class="toc-section-number">3</span> How to build a biomaRt query</a></li>
<li><a href="#examples-of-biomart-queries"><span class="toc-section-number">4</span> Examples of biomaRt queries</a><ul>
<li><a href="#annotate-a-set-of-affymetrix-identifiers-with-hugo-symbol-and-chromosomal-locations-of-corresponding-genes"><span class="toc-section-number">4.1</span> Annotate a set of Affymetrix identifiers with HUGO symbol and chromosomal locations of corresponding genes</a></li>
<li><a href="#annotate-a-set-of-entrezgene-identifiers-with-go-annotation"><span class="toc-section-number">4.2</span> Annotate a set of EntrezGene identifiers with GO annotation</a></li>
<li><a href="#retrieve-all-hugo-gene-symbols-of-genes-that-are-located-on-chromosomes-1720-or-y-and-are-associated-with-specific-go-terms"><span class="toc-section-number">4.3</span> Retrieve all HUGO gene symbols of genes that are located on chromosomes 17,20 or Y, and are associated with specific GO terms</a></li>
<li><a href="#annotate-set-of-idenfiers-with-interpro-protein-domain-identifiers"><span class="toc-section-number">4.4</span> Annotate set of idenfiers with INTERPRO protein domain identifiers</a></li>
<li><a href="#select-all-affymetrix-identifiers-on-the-hgu133plus2-chip-and-ensembl-gene-identifiers-for-genes-located-on-chromosome-16-between-basepair-1100000-and-1250000."><span class="toc-section-number">4.5</span> Select all Affymetrix identifiers on the hgu133plus2 chip and Ensembl gene identifiers for genes located on chromosome 16 between basepair 1100000 and 1250000.</a></li>
<li><a href="#retrieve-all-entrezgene-identifiers-and-hugo-gene-symbols-of-genes-which-have-a-map-kinase-activity-go-term-associated-with-it."><span class="toc-section-number">4.6</span> Retrieve all entrezgene identifiers and HUGO gene symbols of genes which have a MAP kinase activity GO term associated with it.</a></li>
<li><a href="#given-a-set-of-entrezgene-identifiers-retrieve-100bp-upstream-promoter-sequences"><span class="toc-section-number">4.7</span> Given a set of EntrezGene identifiers, retrieve 100bp upstream promoter sequences</a></li>
<li><a href="#retrieve-all-5-utr-sequences-of-all-genes-that-are-located-on-chromosome-3-between-the-positions-185514033-and-185535839"><span class="toc-section-number">4.8</span> Retrieve all 5 UTR sequences of all genes that are located on chromosome 3 between the positions 185,514,033 and 185,535,839</a></li>
<li><a href="#retrieve-protein-sequences-for-a-given-list-of-entrezgene-identifiers"><span class="toc-section-number">4.9</span> Retrieve protein sequences for a given list of EntrezGene identifiers</a></li>
<li><a href="#retrieve-known-snps-located-on-the-human-chromosome-8-between-positions-148350-and-148612"><span class="toc-section-number">4.10</span> Retrieve known SNPs located on the human chromosome 8 between positions 148350 and 148612</a></li>
<li><a href="#given-the-human-gene-tp53-retrieve-the-human-chromosomal-location-of-this-gene-and-also-retrieve-the-chromosomal-location-and-refseq-id-of-its-homolog-in-mouse."><span class="toc-section-number">4.11</span> Given the human gene TP53, retrieve the human chromosomal location of this gene and also retrieve the chromosomal location and RefSeq id of its homolog in mouse.</a></li>
</ul></li>
<li><a href="#using-archived-versions-of-ensembl"><span class="toc-section-number">5</span> Using archived versions of Ensembl</a></li>
<li><a href="#using-a-biomart-other-than-ensembl"><span class="toc-section-number">6</span> Using a BioMart other than Ensembl</a></li>
<li><a href="#biomart-helper-functions"><span class="toc-section-number">7</span> biomaRt helper functions</a><ul>
<li><a href="#exportfasta"><span class="toc-section-number">7.1</span> exportFASTA</a></li>
<li><a href="#finding-out-more-information-on-filters"><span class="toc-section-number">7.2</span> Finding out more information on filters</a><ul>
<li><a href="#filtertype"><span class="toc-section-number">7.2.1</span> filterType</a></li>
<li><a href="#filteroptions"><span class="toc-section-number">7.2.2</span> filterOptions</a></li>
</ul></li>
<li><a href="#attribute-pages"><span class="toc-section-number">7.3</span> Attribute Pages</a></li>
</ul></li>
<li><a href="#local-biomart-databases"><span class="toc-section-number">8</span> Local BioMart databases</a></li>
<li><a href="#using-select"><span class="toc-section-number">9</span> Using <code>select()</code></a></li>
<li><a href="#session-info"><span class="toc-section-number">10</span> Session Info</a></li>
</ul>
</div>

<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>In recent years a wealth of biological data has become available in public data repositories. Easy access to these valuable data resources and firm integration with data analysis is needed for comprehensive bioinformatics data analysis. The <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> package, provides an interface to a growing collection of databases implementing the <a href="http://www.biomart.org">BioMart software suite</a>. The package enables retrieval of large amounts of data in a uniform way without the need to know the underlying database schemas or write complex SQL queries. Examples of BioMart databases are Ensembl, Uniprot and HapMap. These major databases give <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> users direct access to a diverse set of data and enable a wide range of powerful online queries from R.</p>
</div>
<div id="selecting-a-biomart-database-and-dataset" class="section level1">
<h1><span class="header-section-number">2</span> Selecting a BioMart database and dataset</h1>
<p>Every analysis with <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> starts with selecting a BioMart database to use. A first step is to check which BioMart web services are available. The function <code>listMarts()</code> will display all available BioMart web services</p>
<pre class="r"><code>library(&quot;biomaRt&quot;)
listMarts()</code></pre>
<pre><code>##                biomart               version
## 1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 89
## 2   ENSEMBL_MART_MOUSE      Mouse strains 89
## 3     ENSEMBL_MART_SNP  Ensembl Variation 89
## 4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 89</code></pre>
<p>Note: if the function <code>useMart()</code> runs into proxy problems you should set your proxy first before calling any <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> functions.<br />
You can do this using the Sys.putenv command:</p>
<pre class="r"><code>Sys.setenv(&quot;http_proxy&quot; = &quot;http://my.proxy.org:9999&quot;)</code></pre>
<p>Some users have reported that the workaround above does not work, in this case an alternative proxy solution below can be tried:</p>
<pre class="r"><code>options(RCurlOptions = list(proxy=&quot;uscache.kcc.com:80&quot;,proxyuserpwd=&quot;------:-------&quot;))</code></pre>
<p>The <code>useMart()</code> function can now be used to connect to a specified BioMart database, this must be a valid name given by <code>listMarts()</code>. In the next example we choose to query the Ensembl BioMart database.</p>
<pre class="r"><code>ensembl=useMart(&quot;ensembl&quot;)</code></pre>
<p>BioMart databases can contain several datasets, for Ensembl every species is a different dataset. In a next step we look at which datasets are available in the selected BioMart by using the function <code>listDatasets()</code>.</p>
<pre class="r"><code>listDatasets(ensembl)</code></pre>
<pre><code>##                           dataset                                 description                version
## 1         mgallopavo_gene_ensembl                  Turkey genes (Turkey_2.01)            Turkey_2.01
## 2           tguttata_gene_ensembl             Zebra Finch genes (taeGut3.2.4)            taeGut3.2.4
## 3         eeuropaeus_gene_ensembl                    Hedgehog genes (eriEur1)                eriEur1
## 4           pmarinus_gene_ensembl                Lamprey genes (Pmarinus_7.0)           Pmarinus_7.0
## 5  itridecemlineatus_gene_ensembl                    Squirrel genes (spetri2)                spetri2
## 6            gmorhua_gene_ensembl                         Cod genes (gadMor1)                gadMor1
## 7        nleucogenys_gene_ensembl                      Gibbon genes (Nleu1.0)                Nleu1.0
## 8          pcapensis_gene_ensembl                       Hyrax genes (proCap1)                proCap1
## 9          pvampyrus_gene_ensembl                     Megabat genes (pteVam1)                pteVam1
## 10           sscrofa_gene_ensembl                     Pig genes (Sscrofa10.2)            Sscrofa10.2
## 11             mfuro_gene_ensembl                 Ferret genes (MusPutFur1.0)           MusPutFur1.0
## 12          olatipes_gene_ensembl                         Medaka genes (HdrR)                   HdrR
## 13          pformosa_gene_ensembl Amazon molly genes (Poecilia_formosa-5.1.2) Poecilia_formosa-5.1.2
## 14     dnovemcinctus_gene_ensembl                 Armadillo genes (Dasnov3.0)              Dasnov3.0
## 15        choffmanni_gene_ensembl                       Sloth genes (choHof1)                choHof1
## 16        cporcellus_gene_ensembl                  Guinea Pig genes (cavPor3)                cavPor3
## 17          csabaeus_gene_ensembl                Vervet-AGM genes (ChlSab1.1)              ChlSab1.1
## 18     tnigroviridis_gene_ensembl             Tetraodon genes (TETRAODON 8.0)          TETRAODON 8.0
## 19        ogarnettii_gene_ensembl                    Bushbaby genes (OtoGar3)                OtoGar3
## 20         etelfairi_gene_ensembl       Lesser hedgehog tenrec genes (TENREC)                 TENREC
## 21           pabelii_gene_ensembl                     Orangutan genes (PPYG2)                  PPYG2
## 22        ocuniculus_gene_ensembl                    Rabbit genes (OryCun2.0)              OryCun2.0
## 23         lafricana_gene_ensembl                  Elephant genes (Loxafr3.0)              Loxafr3.0
## 24         loculatus_gene_ensembl                 Spotted gar genes (LepOcu1)                LepOcu1
## 25     acarolinensis_gene_ensembl              Anole lizard genes (AnoCar2.0)              AnoCar2.0
## 26            oaries_gene_ensembl                      Sheep genes (Oar_v3.1)               Oar_v3.1
## 27         sharrisii_gene_ensembl      Tasmanian devil genes (Devil_ref v7.0)         Devil_ref v7.0
## 28    aplatyrhynchos_gene_ensembl                   Duck genes (BGI_duck_1.0)           BGI_duck_1.0
## 29        mdomestica_gene_ensembl                     Opossum genes (monDom5)                monDom5
## 30        tbelangeri_gene_ensembl                  Tree Shrew genes (tupBel1)                tupBel1
## 31        gaculeatus_gene_ensembl                Stickleback genes (BROAD S1)               BROAD S1
## 32           btaurus_gene_ensembl                          Cow genes (UMD3.1)                 UMD3.1
## 33         ecaballus_gene_ensembl                     Horse genes (Equ Cab 2)              Equ Cab 2
## 34          ggorilla_gene_ensembl                   Gorilla genes (gorGor3.1)              gorGor3.1
## 35         oanatinus_gene_ensembl                      Platypus genes (OANA5)                  OANA5
## 36          cjacchus_gene_ensembl             Marmoset genes (C_jacchus3.2.1)         C_jacchus3.2.1
## 37            drerio_gene_ensembl                    Zebrafish genes (GRCz10)                 GRCz10
## 38       scerevisiae_gene_ensembl    Saccharomyces cerevisiae genes (R64-1-1)                R64-1-1
## 39        amexicanus_gene_ensembl                 Cave fish genes (AstMex102)              AstMex102
## 40        ttruncatus_gene_ensembl                     Dolphin genes (turTru1)                turTru1
## 41       falbicollis_gene_ensembl               Flycatcher genes (FicAlb_1.4)             FicAlb_1.4
## 42       xtropicalis_gene_ensembl                     Xenopus genes (JGI 4.2)                JGI 4.2
## 43         mmusculus_gene_ensembl                     Mouse genes (GRCm38.p5)              GRCm38.p5
## 44          mmurinus_gene_ensembl                Mouse Lemur genes (Mmur_2.0)               Mmur_2.0
## 45        xmaculatus_gene_ensembl               Platyfish genes (Xipmac4.4.2)            Xipmac4.4.2
## 46     cintestinalis_gene_ensembl                   C.intestinalis genes (KH)                     KH
## 47          hsapiens_gene_ensembl                    Human genes (GRCh38.p10)             GRCh38.p10
## 48          celegans_gene_ensembl     Caenorhabditis elegans genes (WBcel235)               WBcel235
## 49         csyrichta_gene_ensembl                     Tarsier genes (tarSyr1)                tarSyr1
## 50       rnorvegicus_gene_ensembl                        Rat genes (Rnor_6.0)               Rnor_6.0
## 51         psinensis_gene_ensembl Chinese softshell turtle genes (PelSin_1.0)             PelSin_1.0
## 52         trubripes_gene_ensembl                       Fugu genes (FUGU 4.0)               FUGU 4.0
## 53       cfamiliaris_gene_ensembl                       Dog genes (CanFam3.1)              CanFam3.1
## 54          mmulatta_gene_ensembl                  Macaque genes (Mmul_8.0.1)             Mmul_8.0.1
## 55           panubis_gene_ensembl              Olive baboon genes (PapAnu2.0)              PapAnu2.0
## 56         csavignyi_gene_ensembl                 C.savignyi genes (CSAV 2.0)               CSAV 2.0
## 57     dmelanogaster_gene_ensembl                      Fruitfly genes (BDGP6)                  BDGP6
## 58          neugenii_gene_ensembl                    Wallaby genes (Meug_1.0)               Meug_1.0
## 59            fcatus_gene_ensembl                 Cat genes (Felis_catus_6.2)        Felis_catus_6.2
## 60            vpacos_gene_ensembl                      Alpaca genes (vicPac1)                vicPac1
## 61        mlucifugus_gene_ensembl                  Microbat genes (Myoluc2.0)              Myoluc2.0
## 62         oprinceps_gene_ensembl                  Pika genes (OchPri2.0-Ens)          OchPri2.0-Ens
## 63      amelanoleuca_gene_ensembl                       Panda genes (ailMel1)                ailMel1
## 64        lchalumnae_gene_ensembl                  Coelacanth genes (LatCha1)                LatCha1
## 65          saraneus_gene_ensembl                       Shrew genes (sorAra1)                sorAra1
## 66            dordii_gene_ensembl                Kangaroo rat genes (dipOrd1)                dipOrd1
## 67      ptroglodytes_gene_ensembl               Chimpanzee genes (CHIMP2.1.4)             CHIMP2.1.4
## 68           ggallus_gene_ensembl           Chicken genes (Gallus_gallus-5.0)      Gallus_gallus-5.0
## 69        oniloticus_gene_ensembl                   Tilapia genes (Orenil1.0)              Orenil1.0</code></pre>
<p>To select a dataset we can update the <code>Mart</code> object using the function <code>useDataset()</code>. In the example below we choose to use the hsapiens dataset.</p>
<pre class="r"><code>ensembl = useDataset(&quot;hsapiens_gene_ensembl&quot;,mart=ensembl)</code></pre>
<p>Or alternatively if the dataset one wants to use is known in advance, we can select a BioMart database and dataset in one step by:</p>
<pre class="r"><code>ensembl = useMart(&quot;ensembl&quot;,dataset=&quot;hsapiens_gene_ensembl&quot;)</code></pre>
</div>
<div id="how-to-build-a-biomart-query" class="section level1">
<h1><span class="header-section-number">3</span> How to build a biomaRt query</h1>
<p>The <code>getBM()</code> function has three arguments that need to be introduced: filters, attributes and values. <em>Filters</em> define a restriction on the query. For example you want to restrict the output to all genes located on the human X chromosome then the filter <em>chromosome_name</em> can be used with value X. The <code>listFilters()</code> function shows you all available filters in the selected dataset.</p>
<pre class="r"><code>filters = listFilters(ensembl)
filters[1:5,]</code></pre>
<pre><code>##              name              description
## 1 chromosome_name Chromosome/scaffold name
## 2           start                    Start
## 3             end                      End
## 4      band_start               Band Start
## 5        band_end                 Band End</code></pre>
<p><em>Attributes</em> define the values we are interested in to retrieve. For example we want to retrieve the gene symbols or chromosomal coordinates. The <code>listAttributes()</code> function displays all available attributes in the selected dataset.</p>
<pre class="r"><code>attributes = listAttributes(ensembl)
attributes[1:5,]</code></pre>
<pre><code>##                    name          description         page
## 1       ensembl_gene_id       Gene stable ID feature_page
## 2 ensembl_transcript_id Transcript stable ID feature_page
## 3    ensembl_peptide_id    Protein stable ID feature_page
## 4       ensembl_exon_id       Exon stable ID feature_page
## 5           description     Gene description feature_page</code></pre>
<p>The <code>getBM()</code> function is the main query function in <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em>. It has four main arguments:</p>
<ul>
<li><code>attributes</code>: is a vector of attributes that one wants to retrieve (= the output of the query).</li>
<li><code>filters</code>: is a vector of filters that one wil use as input to the query.</li>
<li><code>values</code>: a vector of values for the filters. In case multple filters are in use, the values argument requires a list of values where each position in the list corresponds to the position of the filters in the filters argument (see examples below).</li>
<li><code>mart</code>: is an object of class <code>Mart</code>, which is created by the <code>useMart()</code> function.</li>
</ul>
<p><em>Note: for some frequently used queries to Ensembl, wrapper functions are available: <code>getGene()</code> and <code>getSequence()</code>. These functions call the <code>getBM()</code> function with hard coded filter and attribute names.</em></p>
<p>Now that we selected a BioMart database and dataset, and know about attributes, filters, and the values for filters; we can build a <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> query. Lets make an easy query for the following problem: We have a list of Affymetrix identifiers from the u133plus2 platform and we want to retrieve the corresponding EntrezGene identifiers using the Ensembl mappings.</p>
<p>The u133plus2 platform will be the filter for this query and as values for this filter we use our list of Affymetrix identifiers. As output (attributes) for the query we want to retrieve the EntrezGene and u133plus2 identifiers so we get a mapping of these two identifiers as a result. The exact names that we will have to use to specify the attributes and filters can be retrieved with the <code>listAttributes()</code> and <code>listFilters()</code> function respectively. Lets now run the query:</p>
<pre class="r"><code>affyids=c(&quot;202763_at&quot;,&quot;209310_s_at&quot;,&quot;207500_at&quot;)
getBM(attributes=c('affy_hg_u133_plus_2', 'entrezgene'), 
      filters = 'affy_hg_u133_plus_2', 
      values = affyids, 
      mart = ensembl)</code></pre>
<pre><code>##   affy_hg_u133_plus_2 entrezgene
## 1           202763_at        836
## 2         209310_s_at        837
## 3           207500_at        838</code></pre>
</div>
<div id="examples-of-biomart-queries" class="section level1">
<h1><span class="header-section-number">4</span> Examples of biomaRt queries</h1>
<p>In the sections below a variety of example queries are described. Every example is written as a task, and we have to come up with a <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> solution to the problem.</p>
<div id="annotate-a-set-of-affymetrix-identifiers-with-hugo-symbol-and-chromosomal-locations-of-corresponding-genes" class="section level2">
<h2><span class="header-section-number">4.1</span> Annotate a set of Affymetrix identifiers with HUGO symbol and chromosomal locations of corresponding genes</h2>
<p>We have a list of Affymetrix hgu133plus2 identifiers and we would like to retrieve the HUGO gene symbols, chromosome names, start and end positions and the bands of the corresponding genes. The <code>listAttributes()</code> and the <code>listFilters()</code> functions give us an overview of the available attributes and filters and we look in those lists to find the corresponding attribute and filter names we need. For this query well need the following attributes: hgnc_symbol, chromsome_name, start_position, end_position, band and affy_hg_u133_plus_2 (as we want these in the output to provide a mapping with our original Affymetrix input identifiers. There is one filter in this query which is the affy_hg_u133_plus_2 filter as we use a list of Affymetrix identifiers as input. Putting this all together in the <code>getBM()</code> and performing the query gives:</p>
<pre class="r"><code>affyids=c(&quot;202763_at&quot;,&quot;209310_s_at&quot;,&quot;207500_at&quot;)
getBM(attributes = c('affy_hg_u133_plus_2', 'hgnc_symbol', 'chromosome_name',
                   'start_position', 'end_position', 'band'),
      filters = 'affy_hg_u133_plus_2', 
      values = affyids, 
      mart = ensembl)</code></pre>
<pre><code>##   affy_hg_u133_plus_2 hgnc_symbol chromosome_name start_position end_position  band
## 1           202763_at       CASP3               4      184627696    184649509 q35.1
## 2         209310_s_at       CASP4              11      104942866    104969436 q22.3
## 3           207500_at       CASP5              11      104994235    105023168 q22.3</code></pre>
</div>
<div id="annotate-a-set-of-entrezgene-identifiers-with-go-annotation" class="section level2">
<h2><span class="header-section-number">4.2</span> Annotate a set of EntrezGene identifiers with GO annotation</h2>
<p>In this task we start out with a list of EntrezGene identiers and we want to retrieve GO identifiers related to biological processes that are associated with these entrezgene identifiers. Again we look at the output of <code>listAttributes()</code> and <code>listFilters()</code> to find the filter and attributes we need. Then we construct the following query:</p>
<pre class="r"><code>entrez=c(&quot;673&quot;,&quot;837&quot;)
goids = getBM(attributes = c('entrezgene', 'go_id'), 
              filters = 'entrezgene', 
              values = entrez, 
              mart = ensembl)
head(goids)</code></pre>
<pre><code>##   entrezgene      go_id
## 1        673 GO:0044297
## 2        673 GO:0043005
## 3        673 GO:0016020
## 4        673 GO:0005886
## 5        673 GO:0005739
## 6        673 GO:0005737</code></pre>
</div>
<div id="retrieve-all-hugo-gene-symbols-of-genes-that-are-located-on-chromosomes-1720-or-y-and-are-associated-with-specific-go-terms" class="section level2">
<h2><span class="header-section-number">4.3</span> Retrieve all HUGO gene symbols of genes that are located on chromosomes 17,20 or Y, and are associated with specific GO terms</h2>
<p>The GO terms we are interested in are: <strong>GO:0051330</strong>, <strong>GO:0000080</strong>, <strong>GO:0000114</strong>, <strong>GO:0000082</strong>. The key to performing this query is to understand that the <code>getBM()</code> function enables you to use more than one filter at the same time. In order to do this, the filter argument should be a vector with the filter names. The values should be a list, where the first element of the list corresponds to the first filter and the second list element to the second filter and so on. The elements of this list are vectors containing the possible values for the corresponding filters.</p>
<pre class="r"><code> go=c(&quot;GO:0051330&quot;,&quot;GO:0000080&quot;,&quot;GO:0000114&quot;,&quot;GO:0000082&quot;)
 chrom=c(17,20,&quot;Y&quot;)
 getBM(attributes= &quot;hgnc_symbol&quot;,
        filters=c(&quot;go_id&quot;,&quot;chromosome_name&quot;),
        values=list(go, chrom), mart=ensembl)</code></pre>
<pre><code>## Error in getBM(attributes = &quot;hgnc_symbol&quot;, filters = c(&quot;go_id&quot;, &quot;chromosome_name&quot;), : Invalid filters(s): go_id 
## Please use the function 'listFilters' to get valid filter names</code></pre>
</div>
<div id="annotate-set-of-idenfiers-with-interpro-protein-domain-identifiers" class="section level2">
<h2><span class="header-section-number">4.4</span> Annotate set of idenfiers with INTERPRO protein domain identifiers</h2>
<p>In this example we want to annotate the following two RefSeq identifiers: <strong>NM_005359</strong> and <strong>NM_000546</strong> with INTERPRO protein domain identifiers and a description of the protein domains.</p>
<pre class="r"><code>refseqids = c(&quot;NM_005359&quot;,&quot;NM_000546&quot;)
ipro = getBM(attributes=c(&quot;refseq_mrna&quot;,&quot;interpro&quot;,&quot;interpro_description&quot;), 
             filters=&quot;refseq_mrna&quot;,
             values=refseqids, 
             mart=ensembl)
ipro</code></pre>
<pre><code>##    refseq_mrna  interpro                                   interpro_description
## 1    NM_000546 IPR002117                           p53 tumour suppressor family
## 2    NM_000546 IPR008967             p53-like transcription factor, DNA-binding
## 3    NM_000546 IPR010991                            p53, tetramerisation domain
## 4    NM_000546 IPR011615                                p53, DNA-binding domain
## 5    NM_000546 IPR012346 p53/RUNT-type transcription factor, DNA-binding domain
## 6    NM_000546 IPR013872                             p53 transactivation domain
## 7    NM_005359 IPR001132                              SMAD domain, Dwarfin-type
## 8    NM_005359 IPR003619                           MAD homology 1, Dwarfin-type
## 9    NM_005359 IPR008984                                        SMAD/FHA domain
## 10   NM_005359 IPR013019                                      MAD homology, MH1
## 11   NM_005359 IPR013790                                                Dwarfin
## 12   NM_005359 IPR017855                                       SMAD domain-like</code></pre>
</div>
<div id="select-all-affymetrix-identifiers-on-the-hgu133plus2-chip-and-ensembl-gene-identifiers-for-genes-located-on-chromosome-16-between-basepair-1100000-and-1250000." class="section level2">
<h2><span class="header-section-number">4.5</span> Select all Affymetrix identifiers on the hgu133plus2 chip and Ensembl gene identifiers for genes located on chromosome 16 between basepair 1100000 and 1250000.</h2>
<p>In this example we will again use multiple filters: <em>chromosome_name</em>, <em>start</em>, and <em>end</em> as we filter on these three conditions. Note that when a chromosome name, a start position and an end position are jointly used as filters, the BioMart webservice interprets this as return everything from the given chromosome between the given start and end positions.</p>
<pre class="r"><code>getBM(attributes = c('affy_hg_u133_plus_2','ensembl_gene_id'), 
      filters = c('chromosome_name','start','end'),
      values = list(16,1100000,1250000), 
      mart = ensembl)</code></pre>
<pre><code>##    affy_hg_u133_plus_2 ensembl_gene_id
## 1                      ENSG00000260702
## 2            215502_at ENSG00000260532
## 3                      ENSG00000273551
## 4            205845_at ENSG00000196557
## 5                      ENSG00000196557
## 6                      ENSG00000260403
## 7                      ENSG00000259910
## 8                      ENSG00000261294
## 9          220339_s_at ENSG00000116176
## 10                     ENSG00000277010
## 11         205683_x_at ENSG00000197253
## 12         207134_x_at ENSG00000197253
## 13         217023_x_at ENSG00000197253
## 14         210084_x_at ENSG00000197253
## 15         215382_x_at ENSG00000197253
## 16         216474_x_at ENSG00000197253
## 17         205683_x_at ENSG00000172236
## 18         207134_x_at ENSG00000172236
## 19         217023_x_at ENSG00000172236
## 20         210084_x_at ENSG00000172236
## 21         215382_x_at ENSG00000172236
## 22         216474_x_at ENSG00000172236</code></pre>
</div>
<div id="retrieve-all-entrezgene-identifiers-and-hugo-gene-symbols-of-genes-which-have-a-map-kinase-activity-go-term-associated-with-it." class="section level2">
<h2><span class="header-section-number">4.6</span> Retrieve all entrezgene identifiers and HUGO gene symbols of genes which have a MAP kinase activity GO term associated with it.</h2>
<p>The GO identifier for MAP kinase activity is <strong>GO:0004707</strong>. In our query we will use <em>go_id</em> as our filter, and <em>entrezgene</em> and <em>hgnc_symbol</em> as attributes. Heres the query:</p>
<pre class="r"><code>getBM(attributes = c('entrezgene','hgnc_symbol'), 
      filters = 'go', 
      values = 'GO:0004707', 
      mart = ensembl)</code></pre>
<pre><code>##    entrezgene hgnc_symbol
## 1        1432      MAPK14
## 2        5596       MAPK4
## 3      225689      MAPK15
## 4        5603      MAPK13
## 5        5601       MAPK9
## 6       51701         NLK
## 7        5594       MAPK1
## 8        5599       MAPK8
## 9        5602      MAPK10
## 10       6300      MAPK12
## 11       5597       MAPK6
## 12       5600      MAPK11
## 13       5598       MAPK7
## 14       5595       MAPK3</code></pre>
</div>
<div id="given-a-set-of-entrezgene-identifiers-retrieve-100bp-upstream-promoter-sequences" class="section level2">
<h2><span class="header-section-number">4.7</span> Given a set of EntrezGene identifiers, retrieve 100bp upstream promoter sequences</h2>
<p>All sequence related queries to Ensembl are available through the <code>getSequence()</code> wrapper function. <code>getBM()</code> can also be used directly to retrieve sequences but this can get complicated so using getSequence is recommended.</p>
<p>Sequences can be retrieved using the <code>getSequence()</code> function either starting from chromosomal coordinates or identifiers.<br />
The chromosome name can be specified using the <em>chromosome</em> argument. The <em>start</em> and <em>end</em> arguments are used to specify <em>start</em> and <em>end</em> positions on the chromosome. The type of sequence returned can be specified by the <em>seqType</em> argument which takes the following values:</p>
<ul>
<li><em>cdna</em></li>
<li><em>peptide</em> for protein sequences</li>
<li><em>3utr</em> for 3 UTR sequences</li>
<li><em>5utr</em> for 5 UTR sequences</li>
<li><em>gene_exon</em> for exon sequences only</li>
<li><em>transcript_exon</em> for transcript specific exonic sequences only</li>
<li><em>transcript_exon_intron</em> gives the full unspliced transcript, that is exons + introns</li>
<li><em>gene_exon_intron</em> gives the exons + introns of a gene</li>
<li><em>coding</em> gives the coding sequence only</li>
<li><em>coding_transcript_flank</em> gives the flanking region of the transcript including the UTRs, this must be accompanied with a given value for the upstream or downstream attribute</li>
<li><em>coding_gene_flank</em> gives the flanking region of the gene including the UTRs, this must be accompanied with a given value for the upstream or downstream attribute</li>
<li><em>transcript_flank</em> gives the flanking region of the transcript exculding the UTRs, this must be accompanied with a given value for the upstream or downstream attribute</li>
<li><em>gene_flank</em> gives the flanking region of the gene excluding the UTRs, this must be accompanied with a given value for the upstream or downstream attribute</li>
</ul>
<p>In MySQL mode the <code>getSequence()</code> function is more limited and the sequence that is returned is the 5 to 3+ strand of the genomic sequence, given a chromosome, as start and an end position.</p>
<p>This task requires us to retrieve 100bp upstream promoter sequences from a set of EntrzGene identifiers. The <em>type</em> argument in <code>getSequence()</code> can be thought of as the filter in this query and uses the same input names given by <code>listFilters()</code>. In our query we use entrezgene for the type argument. Next we have to specify which type of sequences we want to retrieve, here we are interested in the sequences of the promoter region, starting right next to the coding start of the gene. Setting the <em>seqType</em> to coding_gene_flank will give us what we need. The <em>upstream</em> argument is used to specify how many bp of upstream sequence we want to retrieve, here well retrieve a rather short sequence of 100bp. Putting this all together in <code>getSequence()</code> gives:</p>
<pre class="r"><code>entrez=c(&quot;673&quot;,&quot;7157&quot;,&quot;837&quot;)
getSequence(id = entrez, 
            type=&quot;entrezgene&quot;,
            seqType=&quot;coding_gene_flank&quot;,
            upstream=100, 
            mart=ensembl) </code></pre>
<pre><code>##                                                                                      coding_gene_flank entrezgene
## 1 CCTCCGCCTCCGCCTCCGCCTCCGCCTCCCCCAGCTCTCCGCCTCCCTTCCCCCTCCCCGCCCGACAGCGGCCGCTCGGGCCCCGGCTCTCGGTTATAAG        673
## 2 CACGTTTCCGCCCTTTGCAATAAGGAAATACATAGTTTACTTTCATTTTTGACTCTGAGGCTCTTTCCAACGCTGTAAAAAAGGACAGAGGCTGTTCCCT        837
## 3 TCCTTCTCTGCAGGCCCAGGTGACCCAGGGTTGGAAGTGTCTCATGCTGGATCCCCACTTTTCCTCTTGCAGCAGCCAGACTGCCTTCCGGGTCACTGCC       7157</code></pre>
</div>
<div id="retrieve-all-5-utr-sequences-of-all-genes-that-are-located-on-chromosome-3-between-the-positions-185514033-and-185535839" class="section level2">
<h2><span class="header-section-number">4.8</span> Retrieve all 5 UTR sequences of all genes that are located on chromosome 3 between the positions 185,514,033 and 185,535,839</h2>
<p>As described in the provious task getSequence can also use chromosomal coordinates to retrieve sequences of all genes that lie in the given region. We also have to specify which type of identifier we want to retrieve together with the sequences, here we choose for entrezgene identifiers.</p>
<pre class="r"><code>utr5 = getSequence(chromosome=3, start=185514033, end=185535839,
                   type=&quot;entrezgene&quot;,
                   seqType=&quot;5utr&quot;, 
                   mart=ensembl)
utr5</code></pre>
<pre><code>##                                                                                                                                             5utr
## 1 AGTCCCTAGGGAACTTCCTGTTGTCACCACACCTCTGAGTCGTCTGAGCTCACTGTGAGCAAAATCCCACAGTGGAAACTCTTAAGCCTCTGCGAAGTAAATCATTCTTGTGAATGTGACACACGATCTCTCCAGTTTCCAT
## 2                                                        TGAGCAAAATCCCACAGTGGAAACTCTTAAGCCTCTGCGAAGTAAATCATTCTTGTGAATGTGACACACGATCTCTCCAGTTTCCAT
## 3                                                                                                                           Sequence unavailable
## 4                                                                                                        ATTCTTGTGAATGTGACACACGATCTCTCCAGTTTCCAT
##   entrezgene
## 1     200879
## 2     200879
## 3     200879
## 4     200879</code></pre>
</div>
<div id="retrieve-protein-sequences-for-a-given-list-of-entrezgene-identifiers" class="section level2">
<h2><span class="header-section-number">4.9</span> Retrieve protein sequences for a given list of EntrezGene identifiers</h2>
<p>In this task the type argument specifies which type of identifiers we are using. To get an overview of other valid identifier types we refer to the <code>listFilters()</code> function.</p>
<pre class="r"><code>protein = getSequence(id=c(100, 5728),
                      type=&quot;entrezgene&quot;,
                      seqType=&quot;peptide&quot;, 
                      mart=ensembl)
protein</code></pre>
<pre><code>##                                                                                                                                                                                                                                                                                                                                                                                                                peptide
## 1                                                                                                                                             MAQTPAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAEGLLNVIGMDKPLTLPDFLAKFDYYMPAIAGCREAIKRIAYEFVEMKAKEGVVYVEVRYSPHLLANSKVEPIPWNQAEGDLTPDEVVALVGQGLQEGERDFGVKARSILCCMRHQPNWSPKVVELCKKYQQQTVVAIDLAGDETIPGSSLLPGHVQAYQEAVKSGIHRTVHAGEVGSAEVVKEAVDILKTERLGHGYHTLEDQALYNRLRQENMHFEAQK*
## 2 MTAIIKEIVSRNKRRYQEDGFDLDLTYIYPNIIAMGFPAERLEGVYRNNIDDVVRFLDSKHKNHYKIYNLCAERHYDTAKFNCRVAQYPFEDHNPPQLELIKPFCEDLDQWLSEDDNHVAAIHCKAGKGRTGVMICAYLLHRGKFLKAQEALDFYGEVRTRDKKGVTIPSQRRYVYYYSYLLKNHLDYRPVALLFHKMMFETIPMFSGGTCNPQFVVCQLKVKIYSSNSGPTRREDKFMYFEFPQPLPVCGDIKVEFFHKQNKMLKKDKMFHFWVNTFFIPGPEETSEKVENGSLCDQEIDSICSIERADNDKEYLVLTLTKNDLDKANKDKANRYFSPNFKVKLYFTKTVEEPSNPEASSSTSVTPDVSDNEPDHYRYSDTTDSDPENEPFDEDQHTQITKV*
## 3                                                                                                                                                                                                                                                                                                                                                                                                 Sequence unavailable
## 4                                                                                                                                                                                                                                                                                                                                                                                                 Sequence unavailable
## 5                                                                                                                                                                                                                                                                                                                                         MAQTPAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAEGLLNVIGMDKPLTLPDFLAKFDYYMPAIARL*
## 6                                                                                                                                                                                                                                                           ALLFHKMMFETIPMFSGGTCNPQFVVCQLKVKIYSSNSGPTRREDKFMYFEFPQPLPVCGDIKVEFFHKQNKMLKKDKMFHFWVNTFFIPGPEETSEKVENGSLCDQEIDSICSIERADNDKEYLVLTLTKNDLDKANKDKANRYFSPNFKVS*
## 7                                         MAQTPAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAEGLLNVIGMDKPLTLPDFLAKFDYYMPAIAGCREAIKRIAYEFVEMKAKEGVVYVEVRYSPHLLANSKVEPIPWNQAEGDLTPDEVVALVGQGLQEGERDFGVKARSILCCMRHQPNWSPKVVELCKKYQQQTVVAIDLAGDETIPGSSLLPGHVQAYQEAVKSGIHRTVHAGEVGSAEVVKEAVDILKTERLGHGYHTLEDQALYNRLRQENMHFEICPWSSYLTGAWKPDTEHAVIRLKNDQANYSLNTDDPLIFKSTLDTDYQMTKRDMGFTEEEFKRLNINAAKSSFLPEDEKRELLDLLYKAYGMPPSASAGQNL*
## 8                                                                 MAQTPAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAEGLLNVIGMDKPLTLPDFLAKFDYYMPAIAGCREAIKRIAYEFVEMKAKEGVVYVEVRYSPHLLANSKVEPIPWNQAEGDLTPDEVVALVGQGLQEGERDFGVKARSILCCMRHQPNWSPKVVELCKKYQQQTVVAIDLAGDETIPGSSLLPGHVQAYQAVDILKTERLGHGYHTLEDQALYNRLRQENMHFEICPWSSYLTGAWKPDTEHAVIRLKNDQANYSLNTDDPLIFKSTLDTDYQMTKRDMGFTEEEFKRLNINAAKSSFLPEDEKRELLDLLYKAYGMPPSASAGQNL*
##   entrezgene
## 1        100
## 2       5728
## 3       5728
## 4        100
## 5        100
## 6       5728
## 7        100
## 8        100</code></pre>
</div>
<div id="retrieve-known-snps-located-on-the-human-chromosome-8-between-positions-148350-and-148612" class="section level2">
<h2><span class="header-section-number">4.10</span> Retrieve known SNPs located on the human chromosome 8 between positions 148350 and 148612</h2>
<p>For this example well first have to connect to a different BioMart database, namely snp.</p>
<pre class="r"><code>snpmart = useMart(biomart = &quot;ENSEMBL_MART_SNP&quot;, dataset=&quot;hsapiens_snp&quot;)</code></pre>
<p>The <code>listAttributes()</code> and <code>listFilters()</code> functions give us an overview of the available attributes and filters.<br />
From these we need: <em>refsnp_id</em>, <em>allele</em>, <em>chrom_start</em> and <em>chrom_strand</em> as attributes; and as filters well use: <em>chrom_start</em>, <em>chrom_end</em> and <em>chr_name</em>.<br />
Note that when a chromosome name, a start position and an end position are jointly used as filters, the BioMart webservice interprets this as return everything from the given chromosome between the given start and end positions. Putting our selected attributes and filters into getBM gives:</p>
<pre class="r"><code>getBM(attributes = c('refsnp_id','allele','chrom_start','chrom_strand'), 
      filters = c('chr_name','start','end'), 
      values = list(8,148350,148612), 
      mart = snpmart)</code></pre>
<pre><code>##     refsnp_id allele chrom_start chrom_strand
## 1 rs868546642    A/G      148372            1
## 2 rs547420070    A/C      148373            1
## 3  rs77274555    G/A      148391            1
## 4 rs567299969    T/A      148394            1
## 5 rs368076569    G/A      148407            1
## 6 rs745318437    C/G      148497            1
## 7 rs190721891    C/G      148576            1</code></pre>
</div>
<div id="given-the-human-gene-tp53-retrieve-the-human-chromosomal-location-of-this-gene-and-also-retrieve-the-chromosomal-location-and-refseq-id-of-its-homolog-in-mouse." class="section level2">
<h2><span class="header-section-number">4.11</span> Given the human gene TP53, retrieve the human chromosomal location of this gene and also retrieve the chromosomal location and RefSeq id of its homolog in mouse.</h2>
<p>The <code>getLDS()</code> (Get Linked Dataset) function provides functionality to link 2 BioMart datasets which each other and construct a query over the two datasets. In Ensembl, linking two datasets translates to retrieving homology data across species. The usage of getLDS is very similar to <code>getBM()</code>. The linked dataset is provided by a separate <code>Mart</code> object and one has to specify filters and attributes for the linked dataset. Filters can either be applied to both datasets or to one of the datasets. Use the listFilters and listAttributes functions on both <code>Mart</code> objects to find the filters and attributes for each dataset (species in Ensembl). The attributes and filters of the linked dataset can be specified with the attributesL and filtersL arguments. Entering all this information into <code>getLDS()</code> gives:</p>
<pre class="r"><code>human = useMart(&quot;ensembl&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;)
mouse = useMart(&quot;ensembl&quot;, dataset = &quot;mmusculus_gene_ensembl&quot;)
getLDS(attributes = c(&quot;hgnc_symbol&quot;,&quot;chromosome_name&quot;, &quot;start_position&quot;),
       filters = &quot;hgnc_symbol&quot;, values = &quot;TP53&quot;,mart = human,
      attributesL = c(&quot;refseq_mrna&quot;,&quot;chromosome_name&quot;,&quot;start_position&quot;), martL = mouse)</code></pre>
<pre><code>##   HGNC.symbol Chromosome.scaffold.name Gene.start..bp. RefSeq.mRNA.ID Chromosome.scaffold.name.1 Gene.start..bp..1
## 1        TP53                       17         7661779                                        11          69580359
## 2        TP53                       17         7661779   NM_001127233                         11          69580359
## 3        TP53                       17         7661779      NM_011640                         11          69580359</code></pre>
</div>
</div>
<div id="using-archived-versions-of-ensembl" class="section level1">
<h1><span class="header-section-number">5</span> Using archived versions of Ensembl</h1>
<p>It is possible to query archived versions of Ensembl through <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em>.</p>
<!-- 
######################################
I dont think the archive=TRUE version works anymore.  The registry of archives is out-of-date
and none of them are live now.  Identifying the host is far more reliable
Mike Smith - 20/10/2016
######################################

There are currently two ways to access archived versions.

## Using `archive = TRUE` 
First we list the available Ensembl archives by using the `listMarts()` function and setting the archive attribute to TRUE.
Note that not all archives are available this way and it seems that recently this only gives access to few archives if you don't see the version of the archive you need please look at the 2nd way to access archives.

```r
listMarts(archive = TRUE)
```

```
##                        biomart                     version
## 1              ensembl_mart_51                  Ensembl 51
## 2                  snp_mart_51                      SNP 51
## 3                 vega_mart_51                     Vega 32
## 4              ensembl_mart_50                  Ensembl 50
## 5                  snp_mart_50                      SNP 50
## 6                 vega_mart_50                     Vega 32
## 7              ensembl_mart_49   ENSEMBL GENES 49 (SANGER)
## 8     genomic_features_mart_49            Genomic Features
## 9                  snp_mart_49                         SNP
## 10                vega_mart_49                        Vega
## 11             ensembl_mart_48   ENSEMBL GENES 48 (SANGER)
## 12    genomic_features_mart_48            Genomic Features
## 13                 snp_mart_48                         SNP
## 14                vega_mart_48                        Vega
## 15             ensembl_mart_47   ENSEMBL GENES 47 (SANGER)
## 16    genomic_features_mart_47            Genomic Features
## 17                 snp_mart_47                         SNP
## 18                vega_mart_47                        Vega
## 19    compara_mart_homology_47            Compara homology
## 20 compara_mart_multiple_ga_47 Compara multiple alignments
## 21 compara_mart_pairwise_ga_47 Compara pairwise alignments
## 22             ensembl_mart_46   ENSEMBL GENES 46 (SANGER)
## 23    genomic_features_mart_46            Genomic Features
## 24                 snp_mart_46                         SNP
## 25                vega_mart_46                        Vega
## 26    compara_mart_homology_46            Compara homology
## 27 compara_mart_multiple_ga_46 Compara multiple alignments
## 28 compara_mart_pairwise_ga_46 Compara pairwise alignments
## 29             ensembl_mart_45   ENSEMBL GENES 45 (SANGER)
## 30                 snp_mart_45                         SNP
## 31                vega_mart_45                        Vega
## 32    compara_mart_homology_45            Compara homology
## 33 compara_mart_multiple_ga_45 Compara multiple alignments
## 34 compara_mart_pairwise_ga_45 Compara pairwise alignments
## 35             ensembl_mart_44   ENSEMBL GENES 44 (SANGER)
## 36                 snp_mart_44                         SNP
## 37                vega_mart_44                        Vega
## 38    compara_mart_homology_44            Compara homology
## 39 compara_mart_pairwise_ga_44 Compara pairwise alignments
## 40             ensembl_mart_43   ENSEMBL GENES 43 (SANGER)
## 41                 snp_mart_43                         SNP
## 42                vega_mart_43                        Vega
## 43    compara_mart_homology_43            Compara homology
## 44 compara_mart_pairwise_ga_43 Compara pairwise alignments
```
Next we select the archive we want to use using the `useMart()` function, again setting the archive attribute to TRUE and giving the full name of the BioMart e.g. ensembl_mart_46.


```r
ensembl = useMart("ensembl_mart_46", dataset="hsapiens_gene_ensembl", archive = TRUE)
```

```
## Note: requested host was redirected from www.ensembl.org to http://aug2007.archive.ensembl.org:80/biomart/martservice
## When using archived Ensembl versions this sometimes can result in connecting to a newer version than the intended Ensembl version
## Check your ensembl version using listMarts(mart)
```

```
## Error in checkDataset(dataset = dataset, mart = mart): The given dataset:  hsapiens_gene_ensembl , is not valid.  Correct dataset names can be obtained with the listDatasets() function.
```

If you don't know the dataset you want to use could first connect to the BioMart using `useMart()` and then use the `listDatasets()` function on this object.
After you selected the BioMart database and dataset, queries can be performed in the same way as when using the current BioMart versions.

## Accessing archives through specifying the archive host
-->
<p>Use the <a href="http://www.ensembl.org" class="uri">http://www.ensembl.org</a> website and go down the bottom of the page. Click on view in Archive and select the archive you need. Copy the url and use that url as shown below to connect to the specified BioMart database. The example below shows how to query Ensembl 54.</p>
<pre class="r"><code>listMarts(host='may2009.archive.ensembl.org')</code></pre>
<pre><code>##                biomart              version
## 1 ENSEMBL_MART_ENSEMBL           Ensembl 54
## 2     ENSEMBL_MART_SNP Ensembl Variation 54
## 3    ENSEMBL_MART_VEGA              Vega 35
## 4             REACTOME   Reactome(CSHL US) 
## 5     wormbase_current   WormBase (CSHL US)
## 6                pride       PRIDE (EBI UK)</code></pre>
<pre class="r"><code>ensembl54=useMart(host='may2009.archive.ensembl.org', 
                  biomart='ENSEMBL_MART_ENSEMBL', 
                  dataset='hsapiens_gene_ensembl')</code></pre>
</div>
<div id="using-a-biomart-other-than-ensembl" class="section level1">
<h1><span class="header-section-number">6</span> Using a BioMart other than Ensembl</h1>
<p>To demonstrate the use of the <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> package with non-Ensembl databases the next query is performed using the Wormbase ParaSite BioMart. In this example, we use the <code>listMarts()</code> function to find the name of the available marts, given the URL of Wormbase. We use this to connect to Wormbase BioMart, find and select the gene dataset, and print the first 6 available attributes and filters. Then we use a list of gene names as filter and retrieve associated transcript IDs and the transcript biotype.</p>
<pre class="r"><code>listMarts(host = &quot;parasite.wormbase.org&quot;)</code></pre>
<pre><code>##         biomart       version
## 1 parasite_mart ParaSite Mart</code></pre>
<pre class="r"><code>wormbase = useMart(biomart = &quot;parasite_mart&quot;, host = &quot;parasite.wormbase.org&quot;)
listDatasets(wormbase)</code></pre>
<pre><code>##     dataset         description version
## 1 wbps_gene All Species (WBPS9)       9</code></pre>
<pre class="r"><code>wormbase &lt;- useDataset(mart = wormbase, dataset = &quot;wbps_gene&quot;)
head(listFilters(wormbase))</code></pre>
<pre><code>##                  name     description
## 1     species_id_1010          Genome
## 2 nematode_clade_1010  Nematode Clade
## 3     chromosome_name Chromosome name
## 4               start           Start
## 5                 end             End
## 6              strand          Strand</code></pre>
<pre class="r"><code>head(listAttributes(wormbase))</code></pre>
<pre><code>##                      name        description         page
## 1          species_id_key      Internal Name feature_page
## 2    production_name_1010     Genome project feature_page
## 3       display_name_1010        Genome name feature_page
## 4        taxonomy_id_1010        Taxonomy ID feature_page
## 5 assembly_accession_1010 Assembly accession feature_page
## 6     nematode_clade_1010     Nematode clade feature_page</code></pre>
<pre class="r"><code>getBM(attributes = c(&quot;external_gene_id&quot;, &quot;wbps_transcript_id&quot;, &quot;transcript_biotype&quot;), 
      filters=&quot;gene_name&quot;, 
      values=c(&quot;unc-26&quot;,&quot;his-33&quot;), 
      mart=wormbase)</code></pre>
<pre><code>##   external_gene_id wbps_transcript_id transcript_biotype
## 1           his-33           F17E9.13     protein_coding
## 2           unc-26            JC8.10a     protein_coding
## 3           unc-26            JC8.10b     protein_coding
## 4           unc-26          JC8.10c.1     protein_coding
## 5           unc-26          JC8.10c.2     protein_coding
## 6           unc-26            JC8.10d     protein_coding</code></pre>
</div>
<div id="biomart-helper-functions" class="section level1">
<h1><span class="header-section-number">7</span> biomaRt helper functions</h1>
<p>This section describes a set of <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> helper functions that can be used to export FASTA format sequences, retrieve values for certain filters and exploring the available filters and attributes in a more systematic manner.</p>
<div id="exportfasta" class="section level2">
<h2><span class="header-section-number">7.1</span> exportFASTA</h2>
<p>The data.frames obtained by the getSequence function can be exported to FASTA files using the <code>exportFASTA()</code> function. One has to specify the data.frame to export and the filename using the file argument.</p>
</div>
<div id="finding-out-more-information-on-filters" class="section level2">
<h2><span class="header-section-number">7.2</span> Finding out more information on filters</h2>
<div id="filtertype" class="section level3">
<h3><span class="header-section-number">7.2.1</span> filterType</h3>
<p>Boolean filters need a value TRUE or FALSE in <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em>. Setting the value TRUE will include all information that fulfill the filter requirement. Setting FALSE will exclude the information that fulfills the filter requirement and will return all values that dont fulfill the filter. For most of the filters, their name indicates if the type is a boolean or not and they will usually start with with. However this is not a rule and to make sure you got the type right you can use the function <code>filterType()</code> to investigate the type of the filter you want to use.</p>
<pre class="r"><code>filterType(&quot;with_affy_hg_u133_plus_2&quot;,ensembl)</code></pre>
<pre><code>## [1] &quot;boolean_list&quot;</code></pre>
</div>
<div id="filteroptions" class="section level3">
<h3><span class="header-section-number">7.2.2</span> filterOptions</h3>
<p>Some filters have a limited set of values that can be given to them. To know which values these are one can use the <code>filterOptions()</code> function to retrieve the predetermed values of the respective filter.</p>
<pre class="r"><code>filterOptions(&quot;biotype&quot;,ensembl)</code></pre>
<pre><code>## [1] &quot;[3prime_overlapping_ncRNA,antisense,bidirectional_promoter_lncRNA,IG_C_gene,IG_C_pseudogene,IG_D_gene,IG_J_gene,IG_J_pseudogene,IG_pseudogene,IG_V_gene,IG_V_pseudogene,lincRNA,macro_lncRNA,miRNA,misc_RNA,Mt_rRNA,Mt_tRNA,non_coding,polymorphic_pseudogene,processed_pseudogene,processed_transcript,protein_coding,pseudogene,ribozyme,rRNA,scaRNA,scRNA,sense_intronic,sense_overlapping,snoRNA,snRNA,sRNA,TEC,transcribed_processed_pseudogene,transcribed_unitary_pseudogene,transcribed_unprocessed_pseudogene,TR_C_gene,TR_D_gene,TR_J_gene,TR_J_pseudogene,TR_V_gene,TR_V_pseudogene,unitary_pseudogene,unprocessed_pseudogene,vaultRNA]&quot;</code></pre>
<p>If there are no predetermed values e.g.for the entrezgene filter, then <code>filterOptions()</code> will return the type of filter it is. And most of the times the filter name or its description will suggest what values one case use for the respective filter (e.g.entrezgene filter will work with enterzgene identifiers as values)</p>
</div>
</div>
<div id="attribute-pages" class="section level2">
<h2><span class="header-section-number">7.3</span> Attribute Pages</h2>
<p>For large BioMart databases such as Ensembl, the number of attributes displayed by the <code>listAttributes()</code> function can be very large. In BioMart databases, attributes are put together in pages, such as sequences, features, homologs for Ensembl. An overview of the attributes pages present in the respective BioMart dataset can be obtained with the <code>attributePages()</code> function.</p>
<pre class="r"><code>pages = attributePages(ensembl)
pages</code></pre>
<pre><code>## [1] &quot;feature_page&quot; &quot;structure&quot;    &quot;homologs&quot;     &quot;snp&quot;          &quot;snp_somatic&quot;  &quot;sequences&quot;</code></pre>
<p>To show us a smaller list of attributes which belong to a specific page, we can now specify this in the <code>listAttributes()</code> function. <em>The set of attributes is still quite long, so we use <code>head()</code> to show only the first few items here.</em></p>
<pre class="r"><code>head(listAttributes(ensembl, page=&quot;feature_page&quot;))</code></pre>
<pre><code>##                    name              description         page
## 1       ensembl_gene_id           Gene stable ID feature_page
## 2 ensembl_transcript_id     Transcript stable ID feature_page
## 3    ensembl_peptide_id        Protein stable ID feature_page
## 4       ensembl_exon_id           Exon stable ID feature_page
## 5           description         Gene description feature_page
## 6       chromosome_name Chromosome/scaffold name feature_page</code></pre>
<p>We now get a short list of attributes related to the region where the genes are located.</p>
</div>
</div>
<div id="local-biomart-databases" class="section level1">
<h1><span class="header-section-number">8</span> Local BioMart databases</h1>
<p>The <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> package can be used with a local install of a public BioMart database or a locally developed BioMart database and web service. In order for <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> to recognize the database as a BioMart, make sure that the local database you create has a name conform with <code>database_mart_version</code> where database is the name of the database and version is a version number. No more underscores than the ones showed should be present in this name. A possible name is for example <code>ensemblLocal_mart_46</code>. ## Minimum requirements for local database installation More information on installing a local copy of a BioMart database or develop your own BioMart database and webservice can be found on <a href="http://www.biomart.org" class="uri">http://www.biomart.org</a> Once the local database is installed you can use <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> on this database by:</p>
<pre class="r"><code>listMarts(host=&quot;www.myLocalHost.org&quot;, path=&quot;/myPathToWebservice/martservice&quot;)
mart=useMart(&quot;nameOfMyMart&quot;,dataset=&quot;nameOfMyDataset&quot;,host=&quot;www.myLocalHost.org&quot;, path=&quot;/myPathToWebservice/martservice&quot;)</code></pre>
<p>For more information on how to install a public BioMart database see: http://www.biomart.org/install.html and follow link databases.</p>
</div>
<div id="using-select" class="section level1">
<h1><span class="header-section-number">9</span> Using <code>select()</code></h1>
<p>In order to provide a more consistent interface to all annotations in Bioconductor the <code>select()</code>, <code>columns()</code>, <code>keytypes()</code> and <code>keys()</code> have been implemented to wrap some of the existing functionality above. These methods can be called in the same manner that they are used in other parts of the project except that instead of taking a <code>AnnotationDb</code> derived class they take instead a <code>Mart</code> derived class as their 1st argument. Otherwise usage should be essentially the same. You still use <code>columns()</code> to discover things that can be extracted from a <code>Mart</code>, and <code>keytypes()</code> to discover which things can be used as keys with <code>select()</code>.</p>
<pre class="r"><code>mart &lt;- useMart(dataset=&quot;hsapiens_gene_ensembl&quot;,biomart='ensembl')
head(keytypes(mart), n=3)</code></pre>
<pre><code>## [1] &quot;affy_hc_g110&quot;        &quot;affy_hg_focus&quot;       &quot;affy_hg_u133_plus_2&quot;</code></pre>
<pre class="r"><code>head(columns(mart), n=3)</code></pre>
<pre><code>## [1] &quot;3_utr_end&quot;   &quot;3_utr_end&quot;   &quot;3_utr_start&quot;</code></pre>
<p>And you still can use <code>keys()</code> to extract potential keys, for a particular key type.</p>
<pre class="r"><code>k = keys(mart, keytype=&quot;chromosome_name&quot;)
head(k, n=3)</code></pre>
<pre><code>## [1] &quot;1&quot; &quot;2&quot; &quot;3&quot;</code></pre>
<p>When using <code>keys()</code>, you can even take advantage of the extra arguments that are available for others keys methods.</p>
<pre class="r"><code>k = keys(mart, keytype=&quot;chromosome_name&quot;, pattern=&quot;LRG&quot;)
head(k, n=3)</code></pre>
<pre><code>## character(0)</code></pre>
<p>Unfortunately the <code>keys()</code> method will not work with all key types because they are not all supported.</p>
<p>But you can still use <code>select()</code> here to extract columns of data that match a particular set of keys (this is basically a wrapper for <code>getBM()</code>).</p>
<pre class="r"><code>affy=c(&quot;202763_at&quot;,&quot;209310_s_at&quot;,&quot;207500_at&quot;)
select(mart, keys=affy, columns=c('affy_hg_u133_plus_2','entrezgene'),
  keytype='affy_hg_u133_plus_2')</code></pre>
<pre><code>##   affy_hg_u133_plus_2 entrezgene
## 1           202763_at        836
## 2         209310_s_at        837
## 3           207500_at        838</code></pre>
<p>So why would we want to do this when we already have functions like <code>getBM()</code>? For two reasons: 1) for people who are familiar with select and its helper methods, they can now proceed to use <em><a href="http://bioconductor.org/packages/biomaRt">biomaRt</a></em> making the same kinds of calls that are already familiar to them and 2) because the select method is implemented in many places elsewhere, the fact that these methods are shared allows for more convenient programmatic access of all these resources. An example of a package that takes advantage of this is the <em><a href="http://bioconductor.org/packages/OrganismDbi">OrganismDbi</a></em> package. Where several packages can be accessed as if they were one resource.</p>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">10</span> Session Info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.4.0 (2017-04-21)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows Server 2012 R2 x64 (build 9600)
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=C                           LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
## [4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] biomaRt_2.32.1  BiocStyle_2.4.0
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.11         AnnotationDbi_1.38.1 knitr_1.16           magrittr_1.5         IRanges_2.10.2      
##  [6] BiocGenerics_0.22.0  stringr_1.2.0        tools_3.4.0          parallel_3.4.0       Biobase_2.36.2      
## [11] DBI_0.6-1            htmltools_0.3.6      yaml_2.1.14          rprojroot_1.2        digest_0.6.12       
## [16] S4Vectors_0.14.3     bitops_1.0-6         RCurl_1.95-4.8       memoise_1.1.0        evaluate_0.10       
## [21] RSQLite_1.1-2        rmarkdown_1.5        stringi_1.1.5        compiler_3.4.0       backports_1.1.0     
## [26] stats4_3.4.0         XML_3.98-1.7</code></pre>
<pre class="r"><code>warnings()</code></pre>
<pre><code>## NULL</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
